proto/osmosis/incentives/gauge.proto

```diff
@@ -41,4 +41,57 @@ message LockableDurationsInfo {
     (gogoproto.stdduration) = true,
     (gogoproto.moretags) = "yaml:\"lockable_durations\""
   ];
+}
+
+// HistoricalReward stores the accumulator value for rewards
+message HistoricalReward {
+  // Reward per share for each denom
+  repeated cosmos.base.v1beta1.DecCoin cumulative_reward_ratio = 1 [
+    (gogoproto.nullable) = false,
+    (gogoproto.castrepeated) = "github.com/cosmos/cosmos-sdk/types.DecCoins"
+  ];
+  int64 epoch = 2;
+}
+
+// CurrentReward stores reward at the latest reward for each denom + lockDuration
+message CurrentReward {
+  int64 last_processed_epoch = 1;
+  // total shares locked for the denom + lockDuration
+  cosmos.base.v1beta1.Coin total_shares = 2 [
+    (gogoproto.nullable) = false,
+    (gogoproto.castrepeated) = "github.com/cosmos/cosmos-sdk/types.Coin"
+  ];
+  // total reward that is to be distributed
+  repeated cosmos.base.v1beta1.Coin rewards = 3 [
+    (gogoproto.nullable) = false,
+    (gogoproto.castrepeated) = "github.com/cosmos/cosmos-sdk/types.Coins"
+  ];
+  // denom of share
+  string denom = 4;
+  google.protobuf.Duration lock_duration = 5 [
+    (gogoproto.stdduration) = true,
+    (gogoproto.nullable) = false,
+    (gogoproto.moretags) = "yaml:\"duration\""
+  ];
+}
+
+// LastEligibleEpochByDurationAndDenom is a struct to store the latest epoch for each denom + lock duration combination
+message LastEligibleEpochByDurationAndDenom {
```

Comment:
I don't apriori understand the purpose of this struct. (Why isn't the latest epoch for every denom, lock duration combo the latest)

Reply:
or is this instead for when a user is claiming rewards, what the 'start' epoch is?

If so, is this indexed by user-address, rather than lock ID?

Reply:
Ah, this can be rather confusing. So even though a period lock would have a single lock duration that could be locked, why would there be arrays of denom + lock duration?
Lock duration here does not mean the duration of the lock itself, we're keeping an array of lock duration for keeping track of at which epoch for each lock duration was this lock eligible for reward.